### ISSUE TITLE
Proposal: ban IOUtils.closeWhileHandlingException

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#16026]
IOUtils.closeWhileHandlingException eats all exceptions raised on close which makes it very difficult to track down failures. I've burned up 2 days this month alone just hunting down failures that were hidden by it that were detected very indirectly by things like the file and thread leak detectors.

**Comment by jpountz** [comment]
I guess this method should take a compulsory Throwable parameter and call addSuppressed for on it for every Closeable that can't be closed successfully?

**Comment by nik9000** [comment]
> I guess this method should take a compulsory Throwable parameter and call addSuppressed for on it for every Closeable that can't be closed successfully?

I'd be quite happy with that.

**Comment by mikemccand** [comment]
> I guess this method should take a compulsory Throwable parameter 

+1

**Comment by mikemccand** [comment]
> I guess this method should take a compulsory Throwable parameter

Actually there's some interesting history here ... which I dug up while working on heroic exception handling in the translog (#16117).

Lucene's `IOUtils` APIs used to take a prior `IOException`, and use `addSuppressed` on that if any new exceptions occurred.

But this API was dangerous.  See https://issues.apache.org/jira/browse/LUCENE-5654 for the full story, but the summary is that Lucene had code all over the place like this:

```
     @Override
     public void close() throws IOException {
      IOException priorE = null;
       try {
         super.close();
      } catch (IOException ioe) {
         priorE = ioe;
       } finally {
         IOUtils.closeWhileHandlingException(priorE, writer);
       }
     }
```

And if you stare at the code for long enough you'll realize that if say an OOME is hit in that `super.close`, then `priorE` is passed as `null` to `IOUtils.closeWhileHandlingException` thus possibly ignoring (losing) the OOME and throwing some other odd `IOException` which could look a lot like index corruption.  So we removed all IOUtils APIs that took a prior exception!

That said, I think we could maybe safely add them back, if we took `Throwable` not `IOException` as the prior exception?  I.e. do this instead?

```
     @Override
     public void close() throws IOException {
      Throwable priorThrowable = null;
       try {
         super.close();
      } catch (Throwable t) {
         priorThrowable = t;
       } finally {
         IOUtils.closeWhileHandlingException(priorThrowable, writer);
       }
     }
```

Anyway, I'm not really sure what to do here, and heroic exception handling is hard and I think often not justified (adds non-trivial complexing for what are exceptional cases)... but wanted to at least call attention here to why none of Lucene's `IOUtils` methods take a prior exception anymore...
